src/com/facebook/buck/shell/DefaultWorkerTool.java

```diff
@@ -85,7 +103,47 @@ public int getMaxWorkers() {
   }
 
   @Override
+  public boolean isPersistent() {
+    return isPersistent;
+  }
+
+  @Override
   public ImmutableSortedSet<BuildRule> getRuntimeDeps() {
     return getDeps();
   }
+
+  @Override
+  public HashCode getInstanceKey() {
+    return buildOutputInitializer.getBuildOutput().getRuleKey().getHashCode();
+  }
+
+  @Override
+  public Data initializeFromDisk(OnDiskBuildInfo onDiskBuildInfo) throws IOException {
+    Optional<RuleKey> ruleKey =
+        onDiskBuildInfo.getRuleKey(BuildInfo.METADATA_KEY_FOR_RULE_KEY);
```

Comment:
The `RuleKeyBuilderFactory` classes have a cache of all computed rule keys on them, so it should be cheap. The part I worry about when using `OnDiskBuildInfo` is that it can only be safely called after the rule has been built and the call to it is buried in the call stack of a method that doesn't communicate this requirement in any way (`WorkerMacroArg#getWorkerHash`). But since violating that would result in an exception which should be caught by the tests, it is probably not such a big deal.